src/main/haskell/kore/src/Kore/Step/Function/Registry.hs

```diff
@@ -0,0 +1,124 @@
+{-|
+Module      : Kore.Step.Function.Registry
+Description : Creates a registry of function evaluators
+Copyright   : (c) Runtime Verification, 2018
+License     : UIUC/NCSA
+Maintainer  : virgil.serbanuta@runtimeverification.com
+Stability   : experimental
+Portability : portable
+-}
+module Kore.Step.Function.Registry
+    ( extractEvaluators
+    ) where
+
+import           Control.Monad
+                 ( when )
+import           Data.List
+                 ( groupBy, sortBy )
+import qualified Data.Map as Map
+import           Data.Maybe
+                 ( mapMaybe )
+import           Data.Ord
+                 ( comparing )
+
+import Kore.AST.Common
+import Kore.AST.Kore
+       ( UnifiedPattern, UnifiedSortVariable )
+import Kore.AST.MetaOrObject
+       ( MetaOrObject, asUnified )
+import Kore.AST.PureML
+       ( fromPurePattern )
+import Kore.AST.PureToKore
+       ( patternKoreToPure )
+import Kore.AST.Sentence
+       ( SentenceAxiom (..) )
+import Kore.IndexedModule.IndexedModule
+       ( IndexedModule (..), KoreIndexedModule )
+import Kore.Predicate.Predicate
+       ( wrapPredicate )
+import Kore.Step.AxiomPatterns
+       ( AxiomPattern (..) )
+import Kore.Step.Function.Data
+       ( ApplicationFunctionEvaluator (..),
+       CommonApplicationFunctionEvaluator )
+import Kore.Step.Function.UserDefined
+       ( axiomFunctionEvaluator )
+import Kore.Step.StepperAttributes
+       ( StepperAttributes )
+
+{-|Given a 'MeraOrObject' @level@ and a 'KoreIndexedModule', @extractEvaluators@
+creates a registry mappting function symbol identifiers to their
+corresponding 'CommonApplicationFunctionEvaluator's.
+-}
+extractEvaluators
+    :: MetaOrObject level
+    => level
+    -> KoreIndexedModule StepperAttributes
+    -> Map.Map (Id level) [CommonApplicationFunctionEvaluator level]
+extractEvaluators level indexedModule =
+    Map.fromList (map extractPrefix groupedEvaluators)
+  where
+    extractPrefix []                  = error "unexpected case"
+    extractPrefix ((a, b) : reminder) = (a, b : map snd reminder)
+    groupedEvaluators =
+        groupBy
+            (\ (a, _) (c, _) -> a == c)
+            (sortBy
+                (comparing fst)
+                (mapMaybe
+                    (axiomToIdEvaluatorPair
+                        level
+                    )
+                    (snd <$> indexedModuleAxioms indexedModule))
+            )
+
+axiomToIdEvaluatorPair
+    :: MetaOrObject level
+    => level
+    -> SentenceAxiom UnifiedSortVariable UnifiedPattern Variable
```

Comment:
Do you think it makes sense to call patternToAxiomPattern before calling this, and passing the result here instead of the SentenceAxiom?